Prefix temporary variable names with function id

peterschrammel opened the pull request:
Temporary variable names must be globally unique and
changing the number of temporaries in one function
must not change a variable name in another function.
Otherwise, this has adverse effects on goto-diff.

This PR gets rid of various global temporary
variables counters that do not meet above criteria.
Instead, we now always use the function id to prefix
temporaries to eliminate cross-function effects
and increment the numeric suffix only when the
entire variable name already exists, using the rename 
facility that has already been used for the same purpose 
in other places.

peterschrammel committed:
Temporary vars tests for goto-diff

peterschrammel committed:
More permissive regression tests regarding tmp var suffixes

tautschnig commented on regression/cbmc/pointer-function-parameters-2/test.desc:
```diff
@@ -3,10 +3,10 @@ main.c
-^a=&tmp\$\d+!0, tmp\$\d+=&tmp\$\d+!0, tmp\$\d+=4$
+^a=\(\(signed int \*\*\)NULL\), tmp(|\$\d+)=[^,]*, tmp(|\$\d+)=[^,]*$
+^a=&tmp(|\$\d+)!0, tmp(|\$\d+)=\(\(signed int \*\)NULL\), tmp(|\$\d+)=[^,]*$
+^a=&tmp(|\$\d+)!0, tmp(|\$\d+)=&tmp(|\$\d+)!0, tmp(|\$\d+)=([012356789][0-9]*|4[0-9]+)$
```
I believe this lost the change from 5832ffd19edf0ca0cab8ea25e8ce79474083530d

tautschnig commented on regression/cbmc/Quantifiers-type/test.desc:
```diff
@@ -2,8 +2,8 @@ CORE
 ^\*\* Results:$
-^\[main.assertion.1\] assertion tmp_if_expr\$1: FAILURE$
-^\[main.assertion.2\] assertion tmp_if_expr\$2: SUCCESS$
+^\[main.assertion.1\] assertion tmp_if_expr(|\$\d+): FAILURE$
```
I think `(\$\d+)?` is more idiomatic than `(|...)`

tautschnig reviewed (changes requested):
I believe this is very helpful, thinking about the code that dump-c happily generates. Just one problem (as noted below): 5832ffd19edf0ca0cab8ea25e8ce79474083530d got lost.

peterschrammel commented:
@tautschnig, I've addressed your comments.

NlightNFotis reviewed (approved):
LGTM

peterschrammel committed:
Stabilize clinit wrapper function type parameters

Many accesses to the function type would add an
ID_parameters if it is not there, which can cause
inconsistencies in irep comparisons.

tautschnig commented on src/java_bytecode/java_static_initializers.cpp:
```diff
@@ -114,6 +114,7 @@ static void create_clinit_wrapper_symbols(
   symbolt wrapper_method_symbol;
   code_typet wrapper_method_type;
   wrapper_method_type.return_type() = void_typet();
+  wrapper_method_type.parameters(); // to ensure the property is there
```
Why is this necessary?! The only impact that I can think of is the hash value and comparison between irepts.

peterschrammel replied:
Exactly for that reason.

tautschnig replied:
So what code is it that is impacted by changes in the ordering resulting from those differences here?

peterschrammel replied:
goto-diff

tautschnig replied:
But then the same program should always yield the same `irept` (previously without an `ID_parameters` in `named_sub` and now with such an entry). Hence I'm still surprised that this impacts goto-diff in some way.

peterschrammel replied:
In principle, yes. The root cause is likely an instability elsewhere in the code. I found this issue only after diffing two 200MB irep dumps. We'll have to investigate this further by shrinking the large example that I have. Any similar sources of nondeterminism will cause further trouble in future. For now, this fix will solve the problem for the customer.

tautschnig replied:
You might just want to log this in an issue to ensure immediate progress while not forgetting about the problem. As the issue surfaces in goto-diff, however, which data structure in goto-diff is it that is being traversed and found to be spuriously different?

peterschrammel committed:
Prefix temporary var names with function id

 Temporary variable names must be globally unique and
 changing the number of temporaries in one function
 must not change a variable name in another function.
 Otherwise, this has adverse effects on goto-diff.

 This commit gets rid of various global temporary
 variables counters that do not meet above criteria.
 Instead, we now always use the function id to prefix
 temporaries to eliminate cross-function effects
 and increment the numeric suffix only when the
 entire variable name already exists, using the
 rename facility.
